File: macos/Services/MCP/Tools/Tier2/SearchAcrossTablesTool.swift
// SearchAcrossTablesTool.swift
// Gridex
//
// MCP Tool: Search for a keyword across table names, column names, and column comments.

import Foundation

struct SearchAcrossTablesTool: MCPTool {
    let name = "search_across_tables"
    let description = "Search for a keyword across table names, column names, and column comments. Useful for discovering relevant data."
    let tier = MCPPermissionTier.read

    let inputSchema: [String: Any] = [
        "type": "object",
        "properties": [
            "connection_id": [
                "type": "string",
                "description": "Connection identifier"
            ],
            "keyword": [
                "type": "string",
                "description": "Keyword to search for"
            ]
        ],
        "required": ["connection_id", "keyword"]
    ]

    func execute(params: JSONValue, context: MCPToolContext) async throws -> MCPToolResult {
        let connectionId = try extractConnectionId(from: params)

        guard let keyword = params["keyword"]?.stringValue else {
            throw MCPToolError.invalidParameters("keyword is required")
        }

        let permission = await context.checkPermission(tier: tier, connectionId: connectionId)
        if let error = permission.errorMessage {
            throw MCPToolError.permissionDenied(error)
        }

        let (adapter, _) = try await context.getAdapter(for: connectionId)

        let searchTerm = keyword.lowercased()
        var matches: [[String: Any]] = []

        let tables = try await adapter.listTables(schema: nil)

        for table in tables {
            // Check table name
            if table.name.lowercased().contains(searchTerm) {
                matches.append([
                    "type": "table",
                    "table": table.name,
                    "match": table.name
                ])
            }

            // Check columns
            let tableDesc = try await adapter.describeTable(name: table.name, schema: nil)
            for column in tableDesc.columns {
                if column.name.lowercased().contains(searchTerm) {
                    matches.append([
                        "type": "column",
                        "table": table.name,
                        "column": column.name,
                        "data_type": column.dataType
                    ])
                }

                // Check column comment if available
                if let comment = column.comment, comment.lowercased().contains(searchTerm) {
                    matches.append([
                        "type": "column_comment",
                        "table": table.name,
                        "column": column.name,
                        "comment": comment
                    ])
                }
            }

            // Check table comment if available
            if let comment = tableDesc.comment, comment.lowercased().contains(searchTerm) {
                matches.append([
                    "type": "table_comment",
                    "table": table.name,
                    "comment": comment
                ])
            }
        }

        if matches.isEmpty {
            return MCPToolResult(text: "No matches found for '\(keyword)'.")
        }

        let data = try JSONSerialization.data(withJSONObject: matches, options: [.prettyPrinted, .sortedKeys])
        let json = String(data: data, encoding: .utf8) ?? "[]"

        return MCPToolResult(text: "Found \(matches.count) match(es) for '\(keyword)':\n\(json)")
    }
}

File: windows/Gridex/Services/MCP/Tools/Tier2/SearchAcrossTablesTool.cpp
//
// SearchAcrossTablesTool.cpp
//

#include "SearchAcrossTablesTool.h"
#include "../MCPToolHelpers.h"
#include <algorithm>

namespace DBModels
{
    namespace
    {
        std::wstring toLower(std::wstring s)
        {
            std::transform(s.begin(), s.end(), s.begin(), ::towlower);
            return s;
        }
        bool containsCI(const std::wstring& hay, const std::wstring& needle)
        {
            return toLower(hay).find(toLower(needle)) != std::wstring::npos;
        }
    }

    MCPToolResult SearchAcrossTablesTool::execute(const nlohmann::json& params, MCPToolContext& ctx)
    {
        const auto connId = MCPTool::extractConnectionId(params);

        if (!params.contains("keyword") || !params["keyword"].is_string())
            throw MCPToolError::invalidParameters("keyword is required");
        const auto kwUtf8 = params["keyword"].get<std::string>();
        const std::wstring kwW = MCPToolHelpers::fromUtf8(kwUtf8);

        const std::wstring schemaW = (params.contains("schema") && params["schema"].is_string())
            ? MCPToolHelpers::fromUtf8(params["schema"].get<std::string>()) : std::wstring{};

        const auto perm = ctx.checkPermission(tier(), connId);
        if (const auto* msg = perm.errorMessage())
            throw MCPToolError::permissionDenied(*msg);

        auto [adapter, config] = ctx.getAdapter(connId);

        nlohmann::json matches = nlohmann::json::array();
        for (const auto& t : adapter->listTables(schemaW))
        {
            if (containsCI(t.name, kwW))
                matches.push_back({
                    {"type",  "table"},
                    {"table", MCPToolHelpers::toUtf8(t.name)},
                    {"match", MCPToolHelpers::toUtf8(t.name)}
                });

            if (!t.comment.empty() && containsCI(t.comment, kwW))
                matches.push_back({
                    {"type",    "table_comment"},
                    {"table",   MCPToolHelpers::toUtf8(t.name)},
                    {"comment", MCPToolHelpers::toUtf8(t.comment)}
                });

            // describeTable can be slow per-table, but Tier-2 read
            // sits behind a rate limiter — fine for exploration.
            for (const auto& col : adapter->describeTable(t.name, schemaW))
            {
                if (containsCI(col.name, kwW))
                    matches.push_back({
                        {"type",      "column"},
                        {"table",     MCPToolHelpers::toUtf8(t.name)},
                        {"column",    MCPToolHelpers::toUtf8(col.name)},
                        {"data_type", MCPToolHelpers::toUtf8(col.dataType)}
                    });
                if (!col.comment.empty() && containsCI(col.comment, kwW))
                    matches.push_back({
                        {"type",    "column_comment"},
                        {"table",   MCPToolHelpers::toUtf8(t.name)},
                        {"column",  MCPToolHelpers::toUtf8(col.name)},
                        {"comment", MCPToolHelpers::toUtf8(col.comment)}
                    });
            }
        }

        if (matches.empty())
            return MCPToolResult::text("No matches found for '" + kwUtf8 + "'.");

        return MCPToolResult::text(
            "Found " + std::to_string(matches.size()) +
            " match(es) for '" + kwUtf8 + "':\n" + matches.dump(2));
    }
}
